azure/services/roleassignments/roleassignments.go

```diff
@@ -31,119 +32,111 @@ import (
 	"sigs.k8s.io/cluster-api-provider-azure/util/tele"
 )
 
-const azureBuiltInContributorID = "b24988ac-6180-42a0-ab88-20f7382dd24c"
+const serviceName = "roleassignments"
 
 // RoleAssignmentScope defines the scope interface for a role assignment service.
 type RoleAssignmentScope interface {
 	azure.ClusterDescriber
-	RoleAssignmentSpecs() []azure.RoleAssignmentSpec
+	azure.AsyncStatusUpdater
+	RoleAssignmentSpecs(principalID *string) []azure.ResourceSpecGetter
+	RoleAssignmentResourceType() string
+	MachineName() string
 }
 
 // Service provides operations on Azure resources.
 type Service struct {
 	Scope RoleAssignmentScope
 	client
+	async.Reconciler
 	virtualMachinesClient        virtualmachines.Client
 	virtualMachineScaleSetClient scalesets.Client
 }
 
 // New creates a new service.
 func New(scope RoleAssignmentScope) *Service {
+	client := newClient(scope)
 	return &Service{
 		Scope:                        scope,
-		client:                       newClient(scope),
+		client:                       client,
 		virtualMachinesClient:        virtualmachines.NewClient(scope),
 		virtualMachineScaleSetClient: scalesets.NewClient(scope),
+		Reconciler:                   async.New(scope, client, client),
 	}
 }
 
 // Reconcile creates a role assignment.
 func (s *Service) Reconcile(ctx context.Context) error {
-	ctx, _, done := tele.StartSpanWithLogger(ctx, "roleassignments.Service.Reconcile")
+	ctx, log, done := tele.StartSpanWithLogger(ctx, "roleassignments.Service.Reconcile")
 	defer done()
+	ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultAzureServiceReconcileTimeout)
+	defer cancel()
+	log.V(2).Info("reconciling role assignment")
+
+	var principalID *string
+	resourceType := s.Scope.RoleAssignmentResourceType()
+	switch resourceType {
+	case azure.VirtualMachine:
+		ID, err := s.getVMPrincipalID(ctx)
+		if err != nil {
+			return errors.Wrap(err, "failed to get principal ID for VM")
+		}
+		principalID = ID
+	case azure.VirtualMachineScaleSet:
+		ID, err := s.getVMSSPrincipalID(ctx)
+		if err != nil {
+			return errors.Wrap(err, "failed to get principal ID for VMSS")
+		}
+		principalID = ID
+	default:
+		return errors.Errorf("unexpected resource type %q. Expected one of [%s, %s]", resourceType,
+			azure.VirtualMachine, azure.VirtualMachineScaleSet)
+	}
 
-	for _, roleSpec := range s.Scope.RoleAssignmentSpecs() {
-		switch roleSpec.ResourceType {
-		case azure.VirtualMachine:
-			return s.reconcileVM(ctx, roleSpec)
-		case azure.VirtualMachineScaleSet:
-			return s.reconcileVMSS(ctx, roleSpec)
-		default:
-			return errors.Errorf("unexpected resource type %q. Expected one of [%s, %s]", roleSpec.ResourceType,
-				azure.VirtualMachine, azure.VirtualMachineScaleSet)
+	for _, roleAssignmentSpec := range s.Scope.RoleAssignmentSpecs(principalID) {
```

Comment:
This is looking really good! I like that we don't need to explicitly cast the spec here anymore.